### PR TITLE
Add SchemaField serialization and deserialization.

### LINKED ISSUE #3419: How to store BigQuery schemas for later use.
Is there an easy way to get the schema from a BigQuery table and save it to disk for later use?

For example, say that I wanted to get the schema of a table that currently exists, save it to disk and later create a table with the same schema.

I see that there is such a thing as a SchemaField in the api, but there doesn't seem to be a clean way to serialize it.

### LINKED ISSUE #3723: BigQuery: expose public helper method to convert a list of schema fields to/from a list of schema dictionaries (JSON)
I've received some feedback asking how to take a schema from the client library and save it to a JSON file. One reason to do this is the [`bq` command-line tool](https://cloud.google.com/bigquery/bq-command-line-tool#creatingtablefromfile) accepts a schema file, formatted like

```
[
  {"name": "name", "type": "string", "mode": "required"},
  {"name": "gender", "type": "string", "mode": "nullable"},
  {"name": "count", "type": "integer", "mode": "required"}
]
```

Note: this format is the same as the API representation.

It would be great if our client libraries could read/write in this format.

### PR DESCRIPTION
**Pull request by lukesneeringer**
Fixes #3419.
Fixes #3723.

### CODE REVIEW

File: bigquery/google/cloud/bigquery/schema.py [hunk 1]
```diff
@@ -84,6 +103,28 @@ def fields(self):
...
+        answer = {
+            'mode': self.mode.lower(),
+            'name': self.name,
+            'type': self.field_type.lower(),
+        }
+
+        # If this is a RECORD type, then sub-fields are also included,
+        # add this to the serialized representation.
+        if self.field_type.upper() == 'RECORD':
+            answer['fields'] = [f.to_api_repr() for f in self.fields]
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/schema.py]: s/`f`/`field` (I know you dislike Pylint, but this is a very easy change to make)

File: bigquery/google/cloud/bigquery/schema.py [hunk 2]
```diff
@@ -84,6 +103,28 @@ def fields(self):
...
         return self._fields
 
+    def to_api_repr(self):
+        """Return a dictionary representing this schema field.
+
+        Returns:
+            dict: A dictionary representing the SchemaField in a serialized
+                form.
+        """
+        # Put together the basic representation. See http://bit.ly/2hOAT5u.
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/schema.py]: What is the value of a `bit.ly` link in our codebase?

Comment (lukesneeringer) [inline bigquery/google/cloud/bigquery/schema.py]: The full link is longer than 80 characters.

File: bigquery/google/cloud/bigquery/schema.py (hunk 1 above)
Comment (lukesneeringer) [inline bigquery/google/cloud/bigquery/schema.py]: Not doing that. The line break makes it _less_ readable and what I have is within the style guide.

**Comment by tseaver** [comment]
This PR should have touched `Table._parse_schema_resource` and `Table._build_schema_resource` as well:  those helpers are used for marshalling a *list* of `SchemaField` instances.

**Comment by tseaver** [comment]
It also shouldn't have closed #3419 / #3723, as it doesn't solve the usecase (serialize / deserialize the *whole* schema, not one field).

**Comment by lukesneeringer** [comment]
Ohhhh, I see.

**Pull request merged**